### DISCUSSION TITLE
Variables in templates and scripts

### DISCUSSION THREAD

**Discussion by ephemeris-lappis** [jmstoolbox/jmstoolbox#186] [Q&A]
Hello.

I'm trying to define templates to make easier sending messages, either one by one, or if I can using scripts.

My first question is about defining variables : I've created custom global variables, but how and when can I set their values ?

Next, are variables accepted to set templates properties ? I've tried setting a property value with "${myVar}", but the literal string is sent...

How can I generate random UUID ?

Thanks in advance for you help.

Regards.

**Comment by titou10titou10** [comment]
> My first question is about defining variables : I've created custom global variables, but how and when can I set their values ?

You don't directly set a value to a variable. The goal of variable is to generate random values. The value is computed by JTB when a template is "hydrated" into a real message, either by drag &drop etc or when a script use a template to send messages
For "global variables" in scripts, you can assign a constant value when defining the variable that will be used acress the messages generated by the script
You can check the doc on [using templates and variables](https://github.com/jmstoolbox/jmstoolbox/wiki/1.4-Templates-and-Variables)  

> 
> Next, are variables accepted to set templates properties ? I've tried setting a property value with "${myVar}", but the literal string is sent...

You are right, variable can not currently be used in JMS properties. It could be a nice enhancement, they are not expanded

> How can I generate random UUID ?

There is no "system" variable for that. I will add one in the next version. As a workaround, define 4 variables like this:

    - uuid-p1: kind: string/custom , length: 8  characters: 0123456789ABCDEF 
    - uuid-p2: kind: string/custom , length: 4, characters: 0123456789ABCDEF 
    - uuid-p3: kind: string/custom , length: 4, characters: 0123456789ABCDEF 
    - uuid-p4: kind: string/custom , length: 8, characters: 0123456789ABCDEF 

Then in your template use:

    ${uuid-p1}-${uuid-p2}-${uuid-p3}-${uuid-p4}

You "can" only define 2 variables and use them like the following, but it will not be very "realistic" as the 8-length parts will be the same as for the 4-length parts 

    ${uuid-p1}-${uuid-p2}-${uuid-p2}-${uuid-p1}

**Comment by titou10titou10** [comment]
I've commit the code for a UUID system variable.
you can grab the code here: https://github.com/jmstoolbox/jmstoolbox/actions/runs/12278836194

**Comment by titou10titou10** [comment]
64442df
This has been resolved in commit [64442df](https://github.com/jmstoolbox/jmstoolbox/commit/64442df258b6c8d33cff9f623b0f9c387c535f07).  
Closing this discussion.
